File: otsu-method/otsu_method.cpp
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>

#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

namespace {

struct Options {
    std::filesystem::path input =
        std::filesystem::path(OTSU_SOURCE_DIR) / "boat.jpg";
    std::filesystem::path output =
        std::filesystem::path(OTSU_SOURCE_DIR) / "outputs" /
        "otsu-opencv-cpp.png";
    bool blur = false;
};

Options parseOptions(int argc, char** argv) {
    Options options;
    for (int index = 1; index < argc; ++index) {
        const std::string argument = argv[index];
        if (argument == "--input" && index + 1 < argc) {
            options.input = argv[++index];
        } else if (argument == "--output" && index + 1 < argc) {
            options.output = argv[++index];
        } else if (argument == "--blur") {
            options.blur = true;
        } else {
            throw std::invalid_argument(
                "Usage: otsu_method [--input IMAGE] [--output IMAGE] "
                "[--blur]");
        }
    }
    return options;
}

}  // namespace

int main(int argc, char** argv) {
    try {
        const auto options = parseOptions(argc, argv);
        cv::Mat image =
            cv::imread(options.input.string(), cv::IMREAD_GRAYSCALE);
        if (image.empty()) {
            throw std::runtime_error(
                "Could not read grayscale image: " + options.input.string());
        }

        cv::Mat processed = image;
        if (options.blur) {
            cv::GaussianBlur(image, processed, cv::Size(5, 5), 0.0);
        }

        cv::Mat binary;
        const double selectedThreshold = cv::threshold(
            processed, binary, 0.0, 255.0,
            cv::THRESH_BINARY | cv::THRESH_OTSU);

        if (!options.output.parent_path().empty()) {
            std::filesystem::create_directories(options.output.parent_path());
        }
        if (!cv::imwrite(options.output.string(), binary)) {
            throw std::runtime_error(
                "Could not write output image: " + options.output.string());
        }

        std::cout << "OpenCV Otsu threshold: " << selectedThreshold << '\n'
                  << "Saved binary image: " << options.output << '\n';
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << '\n';
        return 1;
    }
}

File: otsu-method/otsu_method.py
"""OpenCV's Otsu thresholding call with reproducible, headless outputs."""

from __future__ import annotations

import argparse
from pathlib import Path

import cv2
import matplotlib
import numpy as np

matplotlib.use("Agg")
from matplotlib import pyplot as plt  # noqa: E402


SCRIPT_DIR = Path(__file__).resolve().parent
DEFAULT_INPUT = SCRIPT_DIR / "boat.jpg"
DEFAULT_OUTPUT_DIR = SCRIPT_DIR / "outputs"


def call_otsu_threshold(
    img_title: str | Path = DEFAULT_INPUT,
    is_reduce_noise: bool = False,
    output_dir: str | Path | None = None,
) -> tuple[float, np.ndarray]:
    """Apply OpenCV Otsu thresholding and optionally save plots and output."""

    image_path = Path(img_title).expanduser().resolve()
    image = cv2.imread(str(image_path), cv2.IMREAD_GRAYSCALE)
    if image is None:
        raise FileNotFoundError(f"Could not read grayscale image: {image_path}")

    processed = cv2.GaussianBlur(image, (5, 5), 0) if is_reduce_noise else image
    threshold, binary = cv2.threshold(
        processed, 0, 255, cv2.THRESH_BINARY | cv2.THRESH_OTSU
    )

    if output_dir is not None:
        destination = Path(output_dir).expanduser().resolve()
        destination.mkdir(parents=True, exist_ok=True)

        figure, axis = plt.subplots()
        axis.hist(processed.reshape(-1), bins=256, range=(0, 256))
        axis.set(xlabel="Intensity", ylabel="Pixel count")
        figure.tight_layout()
        figure.savefig(destination / "input-histogram.png", dpi=160)
        plt.close(figure)

        figure, axis = plt.subplots()
        axis.hist(binary.reshape(-1), bins=256, range=(0, 256))
        axis.set(xlabel="Intensity", ylabel="Pixel count")
        figure.tight_layout()
        figure.savefig(destination / "binary-histogram.png", dpi=160)
        plt.close(figure)

        output_path = destination / "otsu-opencv.png"
        if not cv2.imwrite(str(output_path), binary):
            raise OSError(f"Could not write output image: {output_path}")

    return threshold, binary


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Apply OpenCV Otsu thresholding.")
    parser.add_argument("--input", type=Path, default=DEFAULT_INPUT)
    parser.add_argument("--output-dir", type=Path, default=DEFAULT_OUTPUT_DIR)
    parser.add_argument("--blur", action="store_true")
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    threshold, _ = call_otsu_threshold(
        args.input, is_reduce_noise=args.blur, output_dir=args.output_dir
    )
    print(f"OpenCV Otsu threshold: {threshold:g}")
    print(f"Saved outputs under: {args.output_dir.expanduser().resolve()}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
